### PR TITLE
Finishing the manipulatons.py Rewrite

### LINKED ISSUE #43: _Sector method can't average reliably for circular regions.
**Describe the bug**
The class `_Sector` in `sasdata/data_util/manipulations.py` runs into issues when called with values for the args `phi_min` and `phi_max` which are 2pi radians apart. This behaviour can be observed by following the steps below:

**To Reproduce**
1. In Sasview, Compute/Plot any 2D data from `example_data/2d_data`, using any model.
2. Right click the plot and select either 'Wedge Averaging in Q' or 'Wedge Averaging in Phi'.
3. Right click the plot again and select 'Edit Slicer Parameters', and set `delta_phi [deg]` to 180.
4. Try moving the central line around, and note the error in the terminal: `ValueError: Average Error: No points inside sector of ROI to average...`

Sometimes the bug doesn't occur immediately, but it will usually show itself after trying a few different `phi` values. If dragging the central line via the GUI, you'll notice that the slicer won't stay put when you release your mouse button.

**Additional context**
I believe the issue begins with lines 875 and 876 of `manipulations.py`. If `self.phi_min` and `self.phi_max` are 2pi radians apart, then after being passed through the `flip_phi()` function the new variables `phi_min` and `phi_max` have the same value. There are then a load of checks for each of the points in the dataset to see if they're in the region enclosed by `phi_min` and `phi_max`. Instead of examining a circular region for valid datapoints, the `_avg` method examines an infinitesimal region which finishes immediately after it starts. Hence the error message: `ValueError: Average Error: No points inside sector of ROI to average...`.

### LINKED ISSUE #44: Reworking the Slicer Unit Tests
**Is your feature request related to a problem? Please describe.**
The classes of `sasdata/data_util/manipulations.py` make up the backend to sasview's slicers. The respective unit tests for these classes are found in `test/sasdataloader/utest_averaging.py` and these tests leave much to be desired.

There are two classes with tests in `utest_averaging.py`: `Averaging` and `DataInfoTests`.
The tests in `Averaging` seem largely pointless. For starters only two of the eight slicer averaging methods are tested here. These tests are performed using a 100x100 array of 1s as the test dataset, and the test passes result post-averaging is 1. This does not seem like a good test.
The tests in `DataInfoTests` are better, and it actually tests the full spectrum of slicers. What is not ideal however is that their data is loaded in from an external file, and so are the expected results (in most cases). I have been told this is not good practice, and it would be better to mathematically construct this data within the test itself. These tests should have an expected answer that can be calculated analytically, probably also from within the test.

There are some other issues too:
- `test_sectorphi_full` and `test_sectorq_full` do not actually test an azimuthal range of 2pi like their docstrings claim.
- There are no tests for the other aspects of `manipulations.py` such as the supporting (class agnostic) functions
- There are no tests to check how the averagers respond to bad input parameters.
- There are no tests for the mask related classes, `Ringcut`, `Boxcut` and `Sectorcut`.

**Describe the solution you'd like**
The new unit tests should achieve the following:
1. Test all of the averagers under several sets of conditions, particularly where potential bugs may arise.
2. Use test data constructed within the test file. Should be of a mathematically simple form, with expected results that are easily calculated and analytical in form.
3. Test all general-purpose functions present in `manipulations.py` that are used by the averagers.
4. Test the masks? This is lower priority, as I am much less familiar with how these are supposed to work, and how/whether they're implemented in SasView.

**Additional context**
My end goal is a rework of `manipulations.py`, and in order to do a good job of this I need to know that each of the averagers is working as intended. The current unit tests in `DataInfoTests` could be sufficient for this, but making them more robust would be better in the long term.
By writing new unit tests, I'll also become more familiar with the key operations in each of the averaging processes, and therefore better able to simplify and generalise these processes.

### LINKED ISSUE #46: Plan to Restructure/Rewrite manipulations.py
**Is your feature request related to a problem? Please describe.**
As it stands, the file `sasdata/data_util/manipulations.py` is in a very disorderly state. Many of the variables are badly named, large parts of code duplicates functionality, and several bits of code are either obsolete or are not reached in runtime.
The biggest problem however is that each of the averagers goes about its job in a slightly unique way. This makes any future overhauls (such as fractional binning) exponentially more difficult to implement. The abstract methodology behind each of the slicers is fairly universal however, so there should be plenty of opportunity for generalisation and simplification with some restructuring of the classes.

**Describe the solution you'd like**
Broadly speaking, the data processors in `manipulations.py` fit into two categories: the whole-region type and the directional type. The whole-region type includes `Boxsum` and `Boxavg` currently, but with a successfully generalised method it should be easy to add more of this type with various shaped ROIs (regions of interest). Once it has been determined what data is inside the ROI, the process for calculating sums and averages should be universal. The directional type includes all of the others, which return a Data1D object instead of a numerical result. These have the common feature of averaging the intensity along one coordinate axis, returning data to be plotted against the other. This type can be sub-categorised into ones which specify the ROI limits using Cartesian coordinates (`SlabX` and `SlabY`) and ones which use polar coordinates (all the rest). There is further subdivision of the polar kind into Q averagers and Phi averagers, as well as those which use the whole range of a given coordinate axis, or a user-specified region.

The aim is a class structure which reflects the hierarchy described above, using methods which are common between subclasses wherever possible. The rewrite should keep future overhauls to binning in mind, as binning depends heavily on the ROI geometry.

https://github.com/SasView/sasdata/pull/47#pullrequestreview-3465161794

After merging https://github.com/SasView/sasdata/pull/47, initiate PR in SASview to refactor averaging procedure from manipulations.py and point to the refactored averaging classes in https://github.com/SasView/sasdata/blob/78d1f532d27c08d59159c9c7c3ce9ca738682d14/sasdata/data_util/averaging.py . Once this has happened the code in manipulations.py needs clean up of the depreciated code.

### PR DESCRIPTION
**Pull request by ehewins**
## Description

This pull request covers the last part of my summer project: the `manipulations.py` rewrite. It also includes the new analytical unit tests, from branch `44-new-unit-tests-for-slicer-averagers`, except in this branch they point towards the new manipulations module instead of the old one.

The details of the `manipulations.py` rewrite are as follows:
* Parent classes for Cartesian and Polar regions of interest setup have now been added, drastically reducing the amount of repeated code.
* The directional averagers (SlabX, Ring, SectorQ, etc) now all use a common `DirectionalAverage` class for the bulk of their computation. This too reduces repeated code and increases maintainability.
* The method of weighting datapoints based on their position relative to the ROI (inside or outside) has been retooled to make it easier to implement fractional binning later down the line.
* All the data manipulators require an `nbins` parameter now, whereas before some needed `bin_width` instead.

The main functionality change is how `DirectionalAverage` considers the weights of the datapoints. In both the new and old `manipulations.py` these weights are either 0 or 1, though now it should be a lot easier to implement fractional weights. Previously, the weight assigned to a datapoint was calculated on-the-fly as part of a for loop which ran over each datapoint and was also responsible for part of the averaging process. The new methodology separates out the weighting and averaging processes, creating an `n`x`m` weights matrix where `n` is the number of bins the data is sorted into, and `m` is the number of datapoints. The averaging is then done using matrix multiplication and numpy functions. This new method will certainly be more memory heavy than the old method, and it could be slower for large datasets. My hope is however that the switch from python functions to numpy functions will mitigate this, and the improvement to the quality of results once fractional weights are added should more than justify the resource usage.

Dependencies needed: scipy - needed for new unit tests.
Note that merging this pull request requires that a sister pull request in the sasview repo also be merged: **LINK_HERE**

Fixes #46 (issue announcing rewrite plan) and fixes #44 (issue about dodgy unit tests) and fixes #43 (wedge averaging had issues with full-circle ROIs).

## How Has This Been Tested?

There are two ways this rewrite has been tested. The less rigorous method involved using this branch plus the sasview branch mentioned above to confirm that the plotted results look the same before and after using the same data file, slicer and slicer parameters. The more rigorous method involves the new unit tests. When the file `utest_averaging_analytical.py` from the branch `44-new-unit-tests-for-slicer-averagers` is run, you see that averagers from the previous version of `manipulations.py` all pass these tests. The version of `utest_averaging_analytical.py` included in this branch tests the new averagers, and these also pass the tests. Note there are other minor changes made to the tests for compatibility with the new averagers.

## Work still to be done

At a superficial level, there are some files which need renaming. The new version of `manipulations.py` goes by the name `new_manipulations.py`. Once everyone is satisfied with this new version, it should be renamed to replace the old one. There are references to `new_manipulations` in `utest_averaging_analytical.py` on the sasdata side, as well as in `Plotter2D.py`, `AnnulusSlicer.py`, `BoxSlicer.py`, `BoxSum.py`, `WedgeSlicer.py`, and `SectorSlicer.py` on the sasview side. These should also be changed accordingly.

There is also some more serious work to do before this rewrite can truly be considered complete. For the most part we've already reached feature parity with the old `manipulations.py`, but there are some blind spots which need acknowledging.

1. In the old version of `manipulations.py`, the `_Sector` class had the option to bin the data logarithmically. As far as I can tell this functionality was never called upon by sasview. There may be users who use it in custom scripts however, so it would still be worth implementing here. As a bonus, the feature would now be available to all slicers.
2. Currently, no attention is paid to the data's resolution function. In the old `manipulations.py`, `CircularAverage` and `_Sector` call a function by the name of `get_dq_data`. I don't quite understand the maths, but I get the impression from Lucas that it doesn't function as it should. The last thing needed to complete this rewrite would be proper consideration of the `dqx_data` and `dqy_data` properties of the supplied Data2D object to compute the `dx` property of the returned Data1D object. The maths involved here is a little outside my comfort zone, so I'd be happy to delegate this responsibility to someone else.
3. I'd like to restructure the unit tests file. At the moment the tests are grouped in classes based on the averager they're testing, but if instead they were grouped by the kind of test being performed I could use unittest's setUp and tearDown methods to reduce the amount of repeated code. They also only test the averagers under expected conditions. Although the old tests were no more rigorous, I think it would be good write some tests for edge cases.

## Review Checklist (please remove items if they don't apply):

- [ ] Code has been reviewed
- [ ] Functionality has been tested
- [ ] Windows installer (GH artifact) has been tested (installed and worked) 
- [ ] MacOSX installer (GH artifact) has been tested (installed and worked) 
- [X] User documentation is available and complete (if required)
- [X] Developers documentation is available and complete (if required)
- [X] The introduced changes comply with SasView license (BSD 3-Clause)

### CODE REVIEW

**Review by krzywon — COMMENTED** [review]
Code looks good. Locally, I am seeing two errors during testing related to zero division, but the ci succeeds, so likely a local environment issue. I think replacing the scipy requirement with something already included in the package would be good, but not a show-stopper.

Keeping the existing manipulations file is a good idea now that I've had a minute to think about it, but a deprecation message would be helpful. That is something that can be added later.

File: sasdata/data_util/new_manipulations.py
[the remark this answers is not in the document]
Comment (krzywon) [reply sasdata/data_util/new_manipulations.py:1]: Disregard. See my main note. Keeping this is good for backward compatibility.

File: requirements.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ lxml
 
 # Calculation
 numpy
+scipy
```
[the remark this answers is not in the document]
Comment (krzywon) [reply requirements.txt:7]: Considering the comment from @lucas-wilkins, this will be required in the future.

**Comment by wpotrzebowski** [comment]
@lucas-wilkins will take a look at it

**Review by krzywon — APPROVED** [review]
I think this is ready now, but I made changes, so I would suggest an alternate review prior to merging.

File: sasdata/data_util/manipulations.py [hunk 2]
```diff
@@ -22,9 +22,12 @@
 import math
 import numpy as np
 from typing import Optional, Union
+from warnings import warn
 
 from sasdata.dataloader.data_info import Data1D, Data2D
 
+warn("sasdata.data_util.manipulations is deprecated and replaced by sasdata.data_util.averaging.",
```
[the remark this answers is not in the document]
Comment (krzywon) [reply sasdata/data_util/manipulations.py:29]: I did a search and there are still three classes and one function that have not been ported to averaging.py. These need to be  updated to work with the new ROI classes. I plan to tackle that in the next couple of days.

**Comment by wpotrzebowski** [comment]
@smk78 will try to take a look

**Comment by smk78** [comment]
5) Box Averaging in Qx or Qy
Box Average plot appears. Slicer tool appears. Slicer tool can be manipulated by mouse. Plot updates. Until... you try and manipulate the tool from the Slicer Parameters box. Nothing then happens and the tool stops working.
```
ERROR: Traceback (most recent call last): File "sas\qtgui\Plotting\SlicerModel.py", line 66, in setParamsFromModelItem File "sas\qtgui\Plotting\Slicers\BoxSlicer.py", line 242, in setParams KeyError: 'qy_max'
ERROR: Traceback (most recent call last): File "sas\qtgui\Plotting\SlicerModel.py", line 66, in setParamsFromModelItem File "sas\qtgui\Plotting\Slicers\BoxSlicer.py", line 242, in setParams KeyError: 'qy_max'
```

6) Wedge Averaging in Q
Wedge Average plot appears. Slicer tool appears. Slicer tool can be manipulated by mouse. Plot updates. And... with this slicer you _can_ control the tool from the Slicer Parameters box!

7) Wedge Averaging in Phi
Slicer tool appears but then throws an immediate error:
```
ERROR: Traceback (most recent call last): File "sas\qtgui\Plotting\Plotter2D.py", line 480, in onWedgeAveragingPhi File "sas\qtgui\Plotting\Plotter2D.py", line 399, in setSlicer File "sas\qtgui\Plotting\Slicers\WedgeSlicer.py", line 355, in __init__ File "sas\qtgui\Plotting\Slicers\WedgeSlicer.py", line 79, in __init__ File "sas\qtgui\Plotting\Slicers\WedgeSlicer.py", line 361, in _post_data File "sas\qtgui\Plotting\Slicers\WedgeSlicer.py", line 177, in _post_data File "sasdata\data_util\averaging.py", line 899, in __call__ ValueError: operands could not be broadcast together with shapes (82,) (100,) (82,)
```
This tool is also persistent; ie, once it has appeared, if you change the slicer it is not cleared on the plot.

Also, weirdly, the top 14% of the 2D image gets erased with this slicer!

8) General point
In the right-click context menu the Slicers are called:

image

But in the Slicer Parameters box they are called:

image

And the plot labels are:
_(unable to generate plot for Circular Average)_
SectorQ...
AnnulusPhi...
SlabX...
SlabY...
WedgeQ...
_(unable to generate plot for Wedge Averaging in Phi)_

I think some consistency of naming would be helpful.

**Review by smk78 — CHANGES_REQUESTED** [review]
Have performed functionality review on W10/x64. Unfortunately, there are some issues. Please see comments on this PR.
